### PR TITLE
ARROW-3053: [Python] Add unit test for strided object conversion that was failing in 0.9.0

### PR DESCRIPTION
**Pull request by kszucs**

### CODE REVIEW

**Comment by pitrou** [comment]
I see no fix here. Did you manage to reproduce ARROW-3053? Is it required to go through HDF5?

File: python/pyarrow/tests/test_convert_pandas.py [hunk 1]
```diff
@@ -1403,6 +1403,22 @@ def test_decimal_with_None_infer_type(self):
...
+        path = tmpdir.join('test.h5')
+        data = {
+            'a': {0: 'a'},
+            'b': {0: decimal.Decimal('0.0')}
+        }
+
+        df = pd.DataFrame.from_dict(data)
+        df.to_hdf(path, 'test')
+
+        df = pd.read_hdf(path, 'test')
```
Comment (wesm) [inline python/pyarrow/tests/test_convert_pandas.py]: Is HDF5 needed to repro the problem from 0.9.0?

**Comment by wesm** [comment]
Oops missed @pitrou same comment

**Comment by adshieh** [comment]
The roundtrip through HDF5 is required to reproduce the problem. If you make a copy of the DataFrame after the roundtrip, the problem goes away.

**Comment by pitrou** [comment]
Hmm... and did someone investigate why that was?

**Comment by wesm** [comment]
The segfault occurs here, I'm not sure why yet

```
#3  0x00007fffe3a49834 in arrow::py::NumPyConverter::ConvertDecimals (this=0x7fffffffad58) at ../src/arrow/python/numpy_to_arrow.cc:789
789	      RETURN_NOT_OK(max_decimal_metadata.Update(object));
(gdb) l
784	  internal::DecimalMetadata max_decimal_metadata;
785	  Ndarray1DIndexer<PyObject*> objects(arr_);
786	
787	  if (type_ == NULLPTR) {
788	    for (PyObject* object : objects) {
789	      RETURN_NOT_OK(max_decimal_metadata.Update(object));
790	    }
```

**Comment by wesm** [comment]
This script segfaults for me with 0.9.0 

```
import decimal

import pandas as pd
import pyarrow as pa

import pandas.util.testing as tm

data = {
    'a': {0: 'a'},
    'b': {0: decimal.Decimal('0.0')}
}
df = pd.DataFrame.from_dict(data)

nthreads = 1
table = pa.RecordBatch.from_pandas(df, preserve_index=False,
                                   nthreads=nthreads)

result = table.to_pandas(nthreads=nthreads)

tm.assert_frame_equal(result, df, check_dtype=True)
```

**Comment by wesm** [comment]
It does not fail against master. So at least we should nix all the hdf5 stuff in this patch until we understand the underlying problem

**Comment by wesm** [comment]
I found the problem. The "object : objects` logic is implemented incorrectly for NdarrayIndexer:

```
  int64_t size() const { return PyArray_SIZE(arr_); }

  T* data() const { return data_; }

  T* begin() const { return data(); }
  T* end() const { return begin() + size() * stride_; }
```

In this particular case, the stride is 2, so it is reading past the end of the buffer:

```
(gdb) p objects.size()
$5 = 1
(gdb) p object
$6 = (PyObject *) 0x16532d0
(gdb) p objects.stride_
$7 = 2
```

**Comment by wesm** [comment]
I removed the STL iterator protocol methods from `NdarrayIndexer<T>` for now as it does not work correctly. Removed the hdf5 stuff also

**Comment by wesm** [comment]
FYI @adshieh and others who are interested, the reason why HDFStore is in the mix is because of a bug in HDFStore causing columnar / non-strided columns to be returned as strided. I recently reported this as a bug

https://github.com/pandas-dev/pandas/issues/22073

**Comment by wesm** [comment]
+1

**Pull request closed without merging**